lib/TPP/GPU/LinalgToXeGPU.cpp

```diff
@@ -0,0 +1,1394 @@
+  int64_t loadCols = std::min(loadShape2D[1], maxWidth);
+  int64_t arrayLength = std::min(maxWidth / loadCols, maxArrayLength);
+  // In case of partial fit, load only single tile.
+  if (maxWidth % loadCols != 0 || arrayLength != 4 || arrayLength != 2)
+    arrayLength = 1;
+
+  // TODO: Bump IMEX version to bring array_length attr support.
+  arrayLength = 1;
+
+  return createDescriptorTiles(rewriter, loc, src, loadShape2D, {0, 0},
+                               {loadRows, loadCols}, arrayLength);
+}
+
+// Return vector type with specified VNNI shape.
+static VectorType getVnniVector(ArrayRef<int64_t> shape, Type elementType,
+                                VnniConfig vnniConf) {
+  assert(shape.size() == 2 && "Expected plain 2D shape");
+  SmallVector<int64_t> vecShape{shape};
+  vecShape[vnniConf.vnniAxis] /= vnniConf.vnniFactor;
+  vecShape.push_back(vnniConf.vnniFactor);
+  return VectorType::get(vecShape, elementType);
+}
+
+// Loads n-D tiles from memory to registers.
+static SmallVector<Value>
+loadNdDescTiles(PatternRewriter &rewriter, Location loc, ValueRange loadTiles,
+                xegpu::CachePolicyAttr hint,
+                std::optional<VnniConfig> vnniConf = std::nullopt,
+                DenseI64ArrayAttr transpose = nullptr) {
+  // Assume all tiles have the same shape.
+  auto tileType = cast<xegpu::TensorDescType>(loadTiles[0].getType());
+  assert(llvm::all_of(loadTiles,
+                      [&](Value tile) { return tile.getType() == tileType; }) &&
+         "All load tiles must have the same type.");
+
+  VectorType vecLoadType =
+      VectorType::get(tileType.getShape(), tileType.getElementType());
+  IntegerAttr vnniAxisAttr = nullptr;
+  if (vnniConf) {
+    vnniAxisAttr = IntegerAttr::get(rewriter.getI64Type(), vnniConf->vnniAxis);
+    vecLoadType = getVnniVector(tileType.getShape(), tileType.getElementType(),
+                                *vnniConf);
+  }
+
+  SmallVector<Value> loadVec;
+  for (auto tile : loadTiles) {
+    auto loadOp = rewriter.create<xegpu::LoadNdOp>(
+        loc, vecLoadType, tile, vnniAxisAttr, transpose,
+        /*l1_hint=*/hint,
+        /*l2_hint=*/hint, /*l3_hint=*/hint);
+    loadVec.push_back(loadOp);
+  }
+  // TODO: Add split over the array_length > 1.
+  //       The split must preserve row-major ordering of the load tiles.
+
+  return loadVec;
+}
+
+// Splits loaded tiles of a larger 2D tile into individual subtiles and places
+// them in their corresponding positions with respect to the original large
+// tile.
+//
+// The loaded tiles must be perfectly divisible by the specified subtiles.
+// Assumes row-major ordering for both the loaded tiles and the original tile.
+//
+// If the loaded tiles use VNNI layout, corresponding VNNI configuration must be
+// provided.
+static TilesArray
+extractVecSubTiles(PatternRewriter &rewriter, Location loc,
+                   ValueRange loadVecTiles, ArrayRef<int64_t> totalTileShape,
```

Comment:
can we use sgTotalTile, loadTile, dpasTile here?  Just to make it easy to understand

Reply:
Updated naming and comments.